src/tests/runner/test_xml_reporter.cpp

```diff
@@ -0,0 +1,308 @@
+/*
+* (C) 2022 Jack Lloyd
+* (C) 2022 René Meusel, Rohde & Schwarz Cybersecurity
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#include "test_xml_reporter.h"
+
+#if defined(BOTAN_TARGET_OS_HAS_FILESYSTEM)
+
+#include <botan/internal/loadstor.h>
+#include <botan/version.h>
+
+#include <iomanip>
+#include <numeric>
+#include <time.h>
+
+namespace Botan_Tests {
+
+namespace {
+
+std::tm* localtime(const time_t* timer, std::tm *buffer)
+   {
+#if defined(BOTAN_BUILD_COMPILER_IS_MSVC) || \
+    defined(BOTAN_TARGET_OS_IS_MINGW)     || \
+    defined(BOTAN_TARGET_OS_IS_CYGWIN)    || \
+    defined(BOTAN_TARGET_OS_IS_WINDOWS)
+   localtime_s(buffer, timer);
+#else
+   localtime_r(timer, buffer);
+#endif
+   return buffer;
+   }
+
+/// formats a given time point in ISO 8601 format (with time zone)
+std::string format(const std::chrono::system_clock::time_point& tp)
+   {
+   auto seconds_since_epoch = std::chrono::system_clock::to_time_t(tp);
+
+   std::ostringstream out;
+   std::tm buffer{};
+   out << std::put_time(localtime(&seconds_since_epoch, &buffer), "%FT%T%z");
+   return out.str();
+   }
+
+std::string format(const std::chrono::nanoseconds& dur)
+   {
+   const float secs = static_cast<float>(dur.count()) / 1000000000;
+
+   std::ostringstream out;
+   out.precision(3);
+   out << std::fixed << secs;
+   return out.str();
+   }
+
+}
+
+XmlReporter::XmlReporter(const Test_Options& opts, std::string output_dir)
+   : Reporter(opts)
+   , m_output_dir(std::move(output_dir))
+   {
+   set_property("timestamp", format(std::chrono::system_clock::now()));
+   auto custom_props = opts.report_properties();
+   for(const auto& prop : custom_props)
+      {
+      set_property(prop.first, prop.second);
+      }
+   }
+
+XmlReporter::~XmlReporter()
+   {
+   if(m_outfile.has_value() && m_outfile->good())
+      {
+      m_outfile->close();
+      }
+   }
+
+void XmlReporter::render() const
+   {
+   BOTAN_STATE_CHECK(m_outfile.has_value() && m_outfile->good());
+
+   render_preamble(m_outfile.value());
+   render_testsuites(m_outfile.value());
+   }
+
+std::string XmlReporter::get_unique_output_filename() const
+   {
+   const uint64_t ts = Botan_Tests::Test::timestamp();
+   std::vector<uint8_t> seed(8);
+   Botan::store_be(ts, seed.data());
+
+   std::stringstream ss;
+   ss << m_output_dir << "/"
+      << "Botan-"
+      << Botan::short_version_string()
+      << "-tests-"
+      << Botan::hex_encode(seed, false)
+      << ".xml";
+
+   return ss.str();
+   }
+
+void XmlReporter::next_run()
+   {
+   if(m_outfile.has_value() && m_outfile->good())
```

Comment:
Why only explicitly close the file if it is good?

Reply:
Only "good" files may be eligible for graceful handling. 🤡

Changed it to `m_outfile->is_open()`

Reply:
You did change a different place inside the `XmlReporter::~XmlReporter()`. Question is also if in the destructor we maybe could drop it completely. Destructor of `std::basic_ofstream` should close the file automatically.
If we do not check if the close is successful we don't get anything from an explicit close.

Reply:
🤦‍♂️

Yes, certainly: We won't need this destructor. Same in the place you commented in the first place. `std::optional<>::reset()` will destruct the wrapped `std::ofstream` if need be and that will 'do the right thing'™.